lib/MadNLPGPU/src/cusolverrf.jl

```diff
@@ -0,0 +1,180 @@
+# MIT License
+
+# Copyright (c) 2020 Exanauts
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+const CuSubVector{T} = SubArray{T, 1, CUDA.CuArray{T, 1, CUDA.Mem.DeviceBuffer}, Tuple{CUDA.CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}, false}
+
+#=
+    cusolverRF
+=#
+
+@kwdef mutable struct RFSolverOptions <: MadNLP.AbstractOptions
+    rf_symbolic_analysis::Symbol = :klu
+    rf_fast_mode::Bool = true
+    rf_pivot_tol::Float64 = 1e-14
+    rf_boost::Float64 = 1e-14
+    rf_factorization_algo::CUSOLVER.cusolverRfFactorization_t = CUSOLVER.CUSOLVERRF_FACTORIZATION_ALG0
+    rf_triangular_solve_algo::CUSOLVER.cusolverRfTriangularSolve_t = CUSOLVER.CUSOLVERRF_TRIANGULAR_SOLVE_ALG2
```

Comment:
@frapac Was there a reason for setting this to be ALG2? ALG1 seems to be more performant for the OPF cases

Reply:
ha, and interestingly, the `synchronize()` issue above only occurs in `ALG1`

Reply:
I'll revert it back to ALG1 for now but this can be discussed further

Reply:
interesting the synchronize issue only occurs in `ALG1`. I used `ALG2` as this one offered better performance when I ran a benchmark, but the benchmark was far from being exhaustive. I am in to switch back to ALG1 if it has better performance.

Reply:
Quick benchmarking of linear solver time (ALG1 vs ALG2):
- case89: 0.254 vs 0.249
- case1354: 1.054 vs 1.186
- case9241: 4.481 vs 6.772
- case13659: 5.916 vs 7.737
- case30000: 18.668 vs 22.475

Overall, ALG1 seems to be the winner